### PR TITLE
Merge adjacent filter rule for optimizer

### LINKED ISSUE #2016: Merge the adjacent filter
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

```sql
 explain select c1, c2 from test where c3 = true and c2 = 0.000001;

+---------------+-------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                |
+---------------+-------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #test.c1, #test.c2                                                                                                      |
|               |   Filter: #test.c3                                                                                                                  |
|               |     Filter: #test.c2 = Float64(0.000001)                                                                                            |
|               |       TableScan: test projection=Some([0, 1, 2]), filters=[#test.c3, #test.c2 = Float64(0.000001)]                                  |
| physical_plan | ProjectionExec: expr=[c1@0 as c1, c2@1 as c2]                                                                                       |
|               |   CoalesceBatchesExec: target_batch_size=4096                                                                                       |
|               |     FilterExec: c3@2                                                                                                                |
|               |       CoalesceBatchesExec: target_batch_size=4096                                                                                   |
|               |         FilterExec: c2@1 = 0.000001                                                                                                 |
|               |           RepartitionExec: partitioning=RoundRobinBatch(8)                                                                          |
|               |             CsvExec: files=[/home/jakevin/code/arrow-datafusion/datafusion/tests/aggregate_simple.csv], has_header=true, limit=None |
|               |                                                                                                                                     |
+---------------+-------------------------------------------------------------------------------------------------------------------------------------+
```
**Describe the solution you'd like**

add `merge` rule to combine the filter.

**Describe alternatives you've considered**

**Additional context**

Some discussion detail in #2015

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #2016 .

```sql
explain select c1, c2 from test where c3 = true and c2 = 0.000001;
```

Before
```sql
+---------------+-------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                |
+---------------+-------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #test.c1, #test.c2                                                                                                      |
|               |   Filter: #test.c3                                                                                                                  |
|               |     Filter: #test.c2 = Float64(0.000001)                                                                                            |
|               |       TableScan: test projection=Some([0, 1, 2]), filters=[#test.c3, #test.c2 = Float64(0.000001)]                                  |
| physical_plan | ProjectionExec: expr=[c1@0 as c1, c2@1 as c2]                                                                                       |
|               |   CoalesceBatchesExec: target_batch_size=4096                                                                                       |
|               |     FilterExec: c3@2                                                                                                                |
|               |       CoalesceBatchesExec: target_batch_size=4096                                                                                   |
|               |         FilterExec: c2@1 = 0.000001                                                                                                 |
|               |           RepartitionExec: partitioning=RoundRobinBatch(8)                                                                          |
|               |             CsvExec: files=[/home/jakevin/code/arrow-datafusion/datafusion/tests/aggregate_simple.csv], has_header=true, limit=None |
|               |                                                                                                                                     |
+---------------+-------------------------------------------------------------------------------------------------------------------------------------+
```

After
```sql
+---------------+---------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                            |
+---------------+---------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #test.c1, #test.c2                                                                                                  |
|               |   Filter: #test.c3 AND #test.c2 = Float64(0.000001)                                                                             |
|               |     TableScan: test projection=Some([0, 1, 2]), filters=[#test.c3, #test.c2 = Float64(0.000001)]                                |
| physical_plan | ProjectionExec: expr=[c1@0 as c1, c2@1 as c2]                                                                                   |
|               |   CoalesceBatchesExec: target_batch_size=4096                                                                                   |
|               |     FilterExec: c3@2 AND c2@1 = 0.000001                                                                                        |
|               |       RepartitionExec: partitioning=RoundRobinBatch(8)                                                                          |
|               |         CsvExec: files=[/home/jakevin/code/arrow-datafusion/datafusion/tests/aggregate_simple.csv], has_header=true, limit=None |
|               |                                                                                                                                 |
+---------------+---------------------------------------------------------------------------------------------------------------------------------+

```

 # Rationale for this change

merge adjacent filter 

# Are there any user-facing changes?
None

### CODE REVIEW

**Comment by jackwener** [comment]
@alamb @Dandandan @liukun4515  PTAL❤😃.

**Comment by xudong963** [comment]
In fact, I doubt if it really has gains, could you please do some benchmarks for the case? Thanks @jackwener

**Comment by jackwener** [comment]
> In fact, I doubt if it really has gains, could you please do some benchmarks for the case? Thanks @jackwener

Let me explain about this rule. It's a common and base optimizer rule. Postgresql, ,m ysql, cockroach all implement this rule.

like pg:

```sql
explain select id, firstname from scientist where id > 3 and lastname = 's';

QUERY PLAN
--
Seq Scan on scientist (cost=0.00..17.20 rows=1 width=72)
Filter: ((id > 3) AND ((lastname)::text = 's'::text))
```

**Comment by jackwener** [comment]
In many time, It doesn't have a big impact on performance. But it can simplify the query plan.

What' more, other plan can benefit from it, like filter reorder, because we didn't traverse the plan node instead of just focus this filter operator. Because the adjacent filter order is exchangeable.

**Comment by jackwener** [comment]
This postgresql example is more directly. 

```sql
explain select id, firstname from 
(select id, firstname, lastname from scientist where id > 3 ) t
 where lastname = 's';

QUERY PLAN
--
Seq Scan on scientist (cost=0.00..17.20 rows=1 width=72)
Filter: ((id > 3) AND ((lastname)::text = 's'::text))
```

**Comment by xudong963** [comment]
Make sense to me, thank you @jackwener

File: datafusion/src/optimizer/merge_adjacent_filter.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+impl OptimizerRule for MergeAdjacentFilter {
+    fn optimize(
+        &self,
+        plan: &LogicalPlan,
+        execution_props: &ExecutionProps,
+    ) -> Result<LogicalPlan> {
+        let new_plan = optimize(plan);
+
+        // Apply the optimization to all inputs of the plan
+        let inputs = &new_plan.inputs();
```
Comment (xudong963) [inline datafusion/src/optimizer/merge_adjacent_filter.rs:77]: The `&` is redundant?

File: datafusion/src/optimizer/merge_adjacent_filter.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        let formatted_plan = format!("{:?}", optimized_plan);
+        assert_eq!(formatted_plan, expected);
+        assert_eq!(plan.schema(), optimized_plan.schema());
+    }
+
+    #[test]
+    fn double_filter() {
+        let table_scan = test_table_scan().unwrap();
+        let plan =
+            LogicalPlanBuilder::from(table_scan.clone())
```
Comment (xudong963) [inline datafusion/src/optimizer/merge_adjacent_filter.rs:115]:
```suggestion
            LogicalPlanBuilder::from(table_scan)
```

File: datafusion/src/optimizer/merge_adjacent_filter.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            \n  Projection: #test.a\
+            \n    TableScan: test projection=None";
+        assert_optimized_plan_eq(&plan, expected);
+    }
+
+    #[test]
+    fn thrice_filter() {
+        let table_scan = test_table_scan().unwrap();
+        let plan =
+            LogicalPlanBuilder::from(table_scan.clone())
```
Comment (xudong963) [inline datafusion/src/optimizer/merge_adjacent_filter.rs:137]:
```suggestion
            LogicalPlanBuilder::from(table_scan)
```

File: datafusion/src/optimizer/merge_adjacent_filter.rs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            \n  Projection: #test.a\
+            \n    TableScan: test projection=None";
+        assert_optimized_plan_eq(&plan, expected);
+    }
+
+    #[test]
+    fn nested_double_filter() {
+        let table_scan = test_table_scan().unwrap();
+        let plan =
+            LogicalPlanBuilder::from(table_scan.clone())
```
Comment (xudong963) [inline datafusion/src/optimizer/merge_adjacent_filter.rs:161]:
```suggestion
            LogicalPlanBuilder::from(table_scan)
```

**Review by xudong963 — APPROVED** [review]
LGTM, thanks again @jackwener. I commented on a few minor flaws :)

**Comment by liukun4515** [comment]
> # Which issue does this PR close?
> Closes #2016 .
> 
> ```sql
> explain select c1, c2 from test where c3 = true and c2 = 0.000001;
> ```
> 
> Before
> 
> ```sql
> +---------------+-------------------------------------------------------------------------------------------------------------------------------------+
> | plan_type     | plan                                                                                                                                |
> +---------------+-------------------------------------------------------------------------------------------------------------------------------------+
> | logical_plan  | Projection: #test.c1, #test.c2                                                                                                      |
> |               |   Filter: #test.c3                                                                                                                  |
> |               |     Filter: #test.c2 = Float64(0.000001)                                                                                            |
> |               |       TableScan: test projection=Some([0, 1, 2]), filters=[#test.c3, #test.c2 = Float64(0.000001)]                                  |
> | physical_plan | ProjectionExec: expr=[c1@0 as c1, c2@1 as c2]                                                                                       |
> |               |   CoalesceBatchesExec: target_batch_size=4096                                                                                       |
> |               |     FilterExec: c3@2                                                                                                                |
> |               |       CoalesceBatchesExec: target_batch_size=4096                                                                                   |
> |               |         FilterExec: c2@1 = 0.000001                                                                                                 |
> |               |           RepartitionExec: partitioning=RoundRobinBatch(8)                                                                          |
> |               |             CsvExec: files=[/home/jakevin/code/arrow-datafusion/datafusion/tests/aggregate_simple.csv], has_header=true, limit=None |
> |               |                                                                                                                                     |
> +---------------+-------------------------------------------------------------------------------------------------------------------------------------+
> ```
> 
> After
> 
> ```sql
> +---------------+---------------------------------------------------------------------------------------------------------------------------------+
> | plan_type     | plan                                                                                                                            |
> +---------------+---------------------------------------------------------------------------------------------------------------------------------+
> | logical_plan  | Projection: #test.c1, #test.c2                                                                                                  |
> |               |   Filter: #test.c3 AND #test.c2 = Float64(0.000001)                                                                             |
> |               |     TableScan: test projection=Some([0, 1, 2]), filters=[#test.c3, #test.c2 = Float64(0.000001)]                                |
> | physical_plan | ProjectionExec: expr=[c1@0 as c1, c2@1 as c2]                                                                                   |
> |               |   CoalesceBatchesExec: target_batch_size=4096                                                                                   |
> |               |     FilterExec: c3@2 AND c2@1 = 0.000001                                                                                        |
> |               |       RepartitionExec: partitioning=RoundRobinBatch(8)                                                                          |
> |               |         CsvExec: files=[/home/jakevin/code/arrow-datafusion/datafusion/tests/aggregate_simple.csv], has_header=true, limit=None |
> |               |                                                                                                                                 |
> +---------------+---------------------------------------------------------------------------------------------------------------------------------+
> ```
> 
> # Rationale for this change
> merge adjacent filter
> 
> # Are there any user-facing changes?
> None

Could you please provide this schema for this table `test`?

File: datafusion/src/optimizer/merge_adjacent_filter.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+use crate::error::Result;
+use crate::logical_plan::{plan::Filter, LogicalPlan};
+use crate::optimizer::optimizer::OptimizerRule;
+
+use super::utils;
+use crate::execution::context::ExecutionProps;
+
+/// Optimization rule that merge adjacent filter
+#[derive(Default)]
+pub struct MergeAdjacentFilter;
```
Comment (yjshen) [inline datafusion/src/optimizer/merge_adjacent_filter.rs:33]: CombineFilters might be a more suitable name. Besides, a plan could be changed so not adjacent at first.

File: datafusion/src/optimizer/merge_adjacent_filter.rs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            match sub_plan {
+                LogicalPlan::Filter(Filter {
+                    predicate: sub_pred,
+                    input: sub_input,
+                }) => LogicalPlan::Filter(Filter {
+                    predicate: Expr::BinaryExpr {
+                        left: Box::new(pred.clone()),
+                        op: Operator::And,
+                        right: Box::new(sub_pred),
+                    },
```
Comment (yjshen) [inline datafusion/src/optimizer/merge_adjacent_filter.rs:58]: We might also need to remove redundant conditions while combining filters. For instance, `filter(a == 1 && b == 1).filter(a == 1 && c == 1)` could be further optimized to `(a == 1 && b == 1 && c == 1)` from `(a == 1 && b == 1) && (a == 1 && c == 1)`

**Review by yjshen — COMMENTED** [review]
Thanks @jackwener for optimizing this. I think we are on the right track currently but could go steps further.

Regarding the current tests, I think we could:
- Have more tests if we want to eliminate unnecessary conditions (as suggested in the code comments). With both conjunctive and disjunctive filter test cases.
- Add SQL tests as well.

File: datafusion/src/optimizer/merge_adjacent_filter.rs (hunk 6 above)
Comment (jackwener) [reply datafusion/src/optimizer/merge_adjacent_filter.rs:58]: Yes, it's great point. But it should be other rule function.

Now there's `eliminate_fliter`. After this PR, I plan to do enhance this rule for conjunction filter and eliminate more redundant expr after this .

**Comment by doki23** [comment]
Sorry for that, but in my opinion, we should make the sql planner produce only one filter when creating the logical plan instead of creating an optimized rule.

**Review by xudong963 — COMMENTED** [review]
@doki23 's comment reminded me, then I tested it with SQL and found that there is no problem as described in this ticket-related issue.

**Comment by xudong963** [comment]
The following is my test code used by sql.

```rust
#[tokio::test]
async fn main() -> Result<()> {
    // create local execution context
    let mut ctx = SessionContext::new();

    // register csv file with the execution context
    ctx.register_csv("test", "tests/aggregate_simple.csv", CsvReadOptions::new())
        .await?;

    // execute the query
    let plan = ctx.create_logical_plan(
        "select c1, c2 from test where c3 = true and c2 = 0.000001",
    )?;

    dbg!(plan);

    Ok(())
}
```
Then I got the plan
```shell
Projection: #test.c1, #test.c2
  Filter: #test.c3 = Boolean(true) AND #test.c2 = Float64(0.000001)
    TableScan: test projection=Non
```

**Comment by jackwener** [comment]
❤😃 Thanks @xudong963, @doki23

File: datafusion/src/optimizer/merge_adjacent_filter.rs (hunk 6 above)
Comment (alamb) [reply datafusion/src/optimizer/merge_adjacent_filter.rs]: FWIW at least some redundant expressions are already removed by the rule in https://github.com/apache/arrow-datafusion/blob/master/datafusion/src/optimizer/simplify_expressions.rs

```sql
❯ create table foo as select * from (values (1, 2, 3), (4, 5, 6), (7, 8, 9)) as sq;
...
-- Note the rendant `column1 = column2` filter is removed in the actual plan
❯ explain select * from foo where (column1 = column2) AND (column1 = column3) AND (column1 = column2);
+---------------+-----------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                    |
+---------------+-----------------------------------------------------------------------------------------+
| logical_plan  | Projection: #foo.column1, #foo.column2, #foo.column3                                    |
|               |   Filter: #foo.column1 = #foo.column2 AND #foo.column1 = #foo.column3                   |
|               |     TableScan: foo projection=Some([0, 1, 2])                                           |
| physical_plan | ProjectionExec: expr=[column1@0 as column1, column2@1 as column2, column3@2 as column3] |
|               |   CoalesceBatchesExec: target_batch_size=4096                                           |
|               |     FilterExec: column1@0 = column2@1 AND column1@0 = column3@2                         |
|               |       RepartitionExec: partitioning=RoundRobinBatch(16)                                 |
|               |         MemoryExec: partitions=1, partition_sizes=[1]                                   |
|               |                                                                                         |
+---------------+-----------------------------------------------------------------------------------------+
2 rows in set. Query took 0.005 seconds.
```

**Comment by alamb** [comment]
For anyone following along, follow on PR is https://github.com/apache/arrow-datafusion/pull/2039  https://github.com/apache/arrow-datafusion/issues/2038

**Pull request closed without merging**